File: apps/api/src/graphql/query/library.rs
//! Library queries for Resonance GraphQL API
//!
//! This module provides queries for browsing the music library:
//! - Artists: List and search artists
//! - Albums: List and search albums
//! - Tracks: List and search tracks

use async_graphql::{Context, Object, Result, ID};
use uuid::Uuid;

use crate::graphql::pagination::{clamp_limit, clamp_offset, MAX_LIMIT, MAX_SEARCH_LIMIT};
use crate::graphql::types::{Album, Artist, Track};
use crate::repositories::{AlbumRepository, ArtistRepository, TrackRepository};

/// Library-related queries for browsing artists, albums, and tracks
#[derive(Default)]
pub struct LibraryQuery;

#[Object]
impl LibraryQuery {
    // ==================== Artist Queries ====================

    /// Get an artist by ID
    async fn artist(&self, ctx: &Context<'_>, id: ID) -> Result<Option<Artist>> {
        let repo = ctx.data::<ArtistRepository>()?;
        let uuid = Uuid::parse_str(&id)?;
        let artist = repo.find_by_id(uuid).await?;
        Ok(artist.map(Artist::from))
    }

    /// List all artists with pagination
    async fn artists(
        &self,
        ctx: &Context<'_>,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Artist>> {
        let repo = ctx.data::<ArtistRepository>()?;
        let artists = repo
            .find_all(clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(artists.into_iter().map(Artist::from).collect())
    }

    /// Search artists by name
    async fn search_artists(
        &self,
        ctx: &Context<'_>,
        query: String,
        #[graphql(default = 20)] limit: i32,
    ) -> Result<Vec<Artist>> {
        let repo = ctx.data::<ArtistRepository>()?;
        let artists = repo
            .search(&query, clamp_limit(limit, MAX_SEARCH_LIMIT))
            .await?;
        Ok(artists.into_iter().map(Artist::from).collect())
    }

    /// Get artists by genre
    async fn artists_by_genre(
        &self,
        ctx: &Context<'_>,
        genre: String,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Artist>> {
        let repo = ctx.data::<ArtistRepository>()?;
        let artists = repo
            .find_by_genre(&genre, clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(artists.into_iter().map(Artist::from).collect())
    }

    // ==================== Album Queries ====================

    /// Get an album by ID
    async fn album(&self, ctx: &Context<'_>, id: ID) -> Result<Option<Album>> {
        let repo = ctx.data::<AlbumRepository>()?;
        let uuid = Uuid::parse_str(&id)?;
        let album = repo.find_by_id(uuid).await?;
        Ok(album.map(Album::from))
    }

    /// List all albums with pagination
    async fn albums(
        &self,
        ctx: &Context<'_>,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Album>> {
        let repo = ctx.data::<AlbumRepository>()?;
        let albums = repo
            .find_all(clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(albums.into_iter().map(Album::from).collect())
    }

    /// Get albums by artist
    async fn albums_by_artist(
        &self,
        ctx: &Context<'_>,
        artist_id: ID,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Album>> {
        let repo = ctx.data::<AlbumRepository>()?;
        let uuid = Uuid::parse_str(&artist_id)?;
        let albums = repo
            .find_by_artist(uuid, clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(albums.into_iter().map(Album::from).collect())
    }

    /// Search albums by title
    async fn search_albums(
        &self,
        ctx: &Context<'_>,
        query: String,
        #[graphql(default = 20)] limit: i32,
    ) -> Result<Vec<Album>> {
        let repo = ctx.data::<AlbumRepository>()?;
        let albums = repo
            .search(&query, clamp_limit(limit, MAX_SEARCH_LIMIT))
            .await?;
        Ok(albums.into_iter().map(Album::from).collect())
    }

    /// Get recently added albums
    async fn recent_albums(
        &self,
        ctx: &Context<'_>,
        #[graphql(default = 20)] limit: i32,
    ) -> Result<Vec<Album>> {
        let repo = ctx.data::<AlbumRepository>()?;
        let albums = repo
            .find_recent(clamp_limit(limit, MAX_SEARCH_LIMIT))
            .await?;
        Ok(albums.into_iter().map(Album::from).collect())
    }

    // ==================== Track Queries ====================

    /// Get a track by ID
    async fn track(&self, ctx: &Context<'_>, id: ID) -> Result<Option<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let uuid = Uuid::parse_str(&id)?;
        let track = repo.find_by_id(uuid).await?;
        Ok(track.map(Track::from))
    }

    /// List all tracks with pagination
    async fn tracks(
        &self,
        ctx: &Context<'_>,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let tracks = repo
            .find_all(clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(tracks.into_iter().map(Track::from).collect())
    }

    /// Get tracks by album
    async fn tracks_by_album(
        &self,
        ctx: &Context<'_>,
        album_id: ID,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let uuid = Uuid::parse_str(&album_id)?;
        let tracks = repo
            .find_by_album_paginated(uuid, clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(tracks.into_iter().map(Track::from).collect())
    }

    /// Get tracks by artist
    async fn tracks_by_artist(
        &self,
        ctx: &Context<'_>,
        artist_id: ID,
        #[graphql(default = 50)] limit: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> Result<Vec<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let uuid = Uuid::parse_str(&artist_id)?;
        let tracks = repo
            .find_by_artist(uuid, clamp_limit(limit, MAX_LIMIT), clamp_offset(offset))
            .await?;
        Ok(tracks.into_iter().map(Track::from).collect())
    }

    /// Search tracks by title
    async fn search_tracks(
        &self,
        ctx: &Context<'_>,
        query: String,
        #[graphql(default = 20)] limit: i32,
    ) -> Result<Vec<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let tracks = repo
            .search(&query, clamp_limit(limit, MAX_SEARCH_LIMIT))
            .await?;
        Ok(tracks.into_iter().map(Track::from).collect())
    }

    /// Get top played tracks globally
    async fn top_tracks(
        &self,
        ctx: &Context<'_>,
        #[graphql(default = 50)] limit: i32,
    ) -> Result<Vec<Track>> {
        let repo = ctx.data::<TrackRepository>()?;
        let tracks = repo.find_top_tracks(clamp_limit(limit, MAX_LIMIT)).await?;
        Ok(tracks.into_iter().map(Track::from).collect())
    }
}

File: apps/web/src/lib/graphql/library.ts
/**
 * GraphQL library queries for Resonance
 *
 * Contains queries for browsing the music library:
 * - Artists: list, search, by ID
 * - Albums: list, search, recent, by ID
 * - Tracks: list, search, top tracks, by ID
 * - Playlists: user playlists, public playlists, by ID
 */

import { gql } from 'graphql-request'

// ============ Artist Queries ============

/**
 * Get a single artist by ID with albums and top tracks
 */
export const ARTIST_QUERY = gql`
  query Artist($id: ID!) {
    artist(id: $id) {
      id
      name
      sortName
      mbid
      biography
      imageUrl
      genres
      albumCount
      createdAt
      updatedAt
      albums(limit: 50) {
        id
        title
        releaseDate
        releaseYear
        albumType
        coverArtUrl
        totalTracks
      }
      topTracks(limit: 10) {
        id
        title
        durationMs
        formattedDuration
        trackNumber
        streamUrl
        album {
          id
          title
          coverArtUrl
        }
      }
    }
  }
`

/**
 * List all artists with pagination
 */
export const ARTISTS_QUERY = gql`
  query Artists($limit: Int, $offset: Int) {
    artists(limit: $limit, offset: $offset) {
      id
      name
      imageUrl
      genres
      albumCount
    }
  }
`

/**
 * Search artists by name
 */
export const SEARCH_ARTISTS_QUERY = gql`
  query SearchArtists($query: String!, $limit: Int) {
    searchArtists(query: $query, limit: $limit) {
      id
      name
      imageUrl
      genres
    }
  }
`

/**
 * Get artists by genre
 */
export const ARTISTS_BY_GENRE_QUERY = gql`
  query ArtistsByGenre($genre: String!, $limit: Int, $offset: Int) {
    artistsByGenre(genre: $genre, limit: $limit, offset: $offset) {
      id
      name
      imageUrl
      genres
      albumCount
    }
  }
`

// ============ Album Queries ============

/**
 * Get a single album by ID with tracks
 */
export const ALBUM_QUERY = gql`
  query Album($id: ID!) {
    album(id: $id) {
      id
      title
      artistId
      mbid
      releaseDate
      releaseYear
      albumType
      genres
      totalTracks
      totalDurationMs
      formattedDuration
      coverArtPath
      coverArtUrl
      coverArtColors {
        primary
        secondary
        accent
        vibrant
        muted
      }
      createdAt
      updatedAt
      artist {
        id
        name
      }
      tracks(limit: 100) {
        id
        title
        trackNumber
        discNumber
        durationMs
        formattedDuration
        fileFormat
        bitRate
        sampleRate
        bitDepth
        isHires
        isLossless
        explicit
        playCount
        streamUrl
      }
    }
  }
`

/**
 * List all albums with pagination
 */
export const ALBUMS_QUERY = gql`
  query Albums($limit: Int, $offset: Int) {
    albums(limit: $limit, offset: $offset) {
      id
      title
      artistId
      releaseYear
      albumType
      coverArtUrl
      artist {
        id
        name
      }
    }
  }
`

/**
 * Get albums by artist
 */
export const ALBUMS_BY_ARTIST_QUERY = gql`
  query AlbumsByArtist($artistId: ID!, $limit: Int, $offset: Int) {
    albumsByArtist(artistId: $artistId, limit: $limit, offset: $offset) {
      id
      title
      releaseYear
      albumType
      coverArtUrl
      totalTracks
    }
  }
`

/**
 * Search albums by title
 */
export const SEARCH_ALBUMS_QUERY = gql`
  query SearchAlbums($query: String!, $limit: Int) {
    searchAlbums(query: $query, limit: $limit) {
      id
      title
      coverArtUrl
      releaseYear
      albumType
      artist {
        id
        name
      }
    }
  }
`

/**
 * Get recently added albums
 */
export const RECENT_ALBUMS_QUERY = gql`
  query RecentAlbums($limit: Int) {
    recentAlbums(limit: $limit) {
      id
      title
      coverArtUrl
      releaseYear
      artist {
        id
        name
      }
    }
  }
`

// ============ Track Queries ============

/**
 * Get a single track by ID
 */
export const TRACK_QUERY = gql`
  query Track($id: ID!) {
    track(id: $id) {
      id
      title
      artistId
      albumId
      mbid
      fileFormat
      durationMs
      formattedDuration
      bitRate
      sampleRate
      bitDepth
      isHires
      isLossless
      trackNumber
      discNumber
      genres
      explicit
      lyrics
      hasSyncedLyrics
      audioFeatures {
        bpm
        key
        mode
        loudness
        energy
        danceability
        valence
        acousticness
        instrumentalness
        speechiness
      }
      aiMood
      aiTags
      aiDescription
      playCount
      skipCount
      lastPlayedAt
      streamUrl
      createdAt
      updatedAt
      album {
        id
        title
        coverArtUrl
      }
      artist {
        id
        name
      }
    }
  }
`

/**
 * List all tracks with pagination
 */
export const TRACKS_QUERY = gql`
  query Tracks($limit: Int, $offset: Int) {
    tracks(limit: $limit, offset: $offset) {
      id
      title
      durationMs
      formattedDuration
      fileFormat
      isHires
      isLossless
      playCount
      streamUrl
      artist {
        id
        name
      }
      album {
        id
        title
        coverArtUrl
      }
    }
  }
`

/**
 * Get tracks by album
 */
export const TRACKS_BY_ALBUM_QUERY = gql`
  query TracksByAlbum($albumId: ID!, $limit: Int, $offset: Int) {
    tracksByAlbum(albumId: $albumId, limit: $limit, offset: $offset) {
      id
      title
      trackNumber
      discNumber
      durationMs
      formattedDuration
      fileFormat
      isHires
      isLossless
      explicit
      streamUrl
    }
  }
`

/**
 * Get tracks by artist
 */
export const TRACKS_BY_ARTIST_QUERY = gql`
  query TracksByArtist($artistId: ID!, $limit: Int, $offset: Int) {
    tracksByArtist(artistId: $artistId, limit: $limit, offset: $offset) {
      id
      title
      durationMs
      formattedDuration
      playCount
      streamUrl
      album {
        id
        title
        coverArtUrl
      }
    }
  }
`

/**
 * Search tracks by title
 */
export const SEARCH_TRACKS_QUERY = gql`
  query SearchTracks($query: String!, $limit: Int) {
    searchTracks(query: $query, limit: $limit) {
      id
      title
      durationMs
      formattedDuration
      streamUrl
      artist {
        id
        name
      }
      album {
        id
        title
        coverArtUrl
      }
    }
  }
`

/**
 * Get top played tracks globally
 */
export const TOP_TRACKS_QUERY = gql`
  query TopTracks($limit: Int) {
    topTracks(limit: $limit) {
      id
      title
      durationMs
      formattedDuration
      playCount
      streamUrl
      artist {
        id
        name
      }
      album {
        id
        title
        coverArtUrl
      }
    }
  }
`

// ============ Playlist Queries ============

/**
 * Get a single playlist by ID with tracks
 */
export const PLAYLIST_QUERY = gql`
  query Playlist($id: ID!) {
    playlist(id: $id) {
      id
      userId
      name
      description
      imageUrl
      isPublic
      isCollaborative
      playlistType
      trackCount
      totalDurationMs
      formattedDuration
      createdAt
      updatedAt
      tracks(limit: 200) {
        position
        addedBy
        addedAt
        track {
          id
          title
          durationMs
          formattedDuration
          fileFormat
          isHires
          isLossless
          streamUrl
          artist {
            id
            name
          }
          album {
            id
            title
            coverArtUrl
          }
        }
      }
    }
  }
`

/**
 * Get playlists owned by the authenticated user
 */
export const MY_PLAYLISTS_QUERY = gql`
  query MyPlaylists($limit: Int, $offset: Int) {
    myPlaylists(limit: $limit, offset: $offset) {
      id
      name
      description
      imageUrl
      playlistType
      trackCount
      totalDurationMs
      formattedDuration
      updatedAt
    }
  }
`

/**
 * Browse public playlists
 */
export const PUBLIC_PLAYLISTS_QUERY = gql`
  query PublicPlaylists($limit: Int, $offset: Int) {
    publicPlaylists(limit: $limit, offset: $offset) {
      id
      name
      description
      imageUrl
      playlistType
      trackCount
    }
  }
`

// ============ Combined Search Query ============

/**
 * Search across all content types at once
 */
export const COMBINED_SEARCH_QUERY = gql`
  query CombinedSearch($query: String!, $limit: Int) {
    searchArtists(query: $query, limit: $limit) {
      id
      name
      imageUrl
      genres
    }
    searchAlbums(query: $query, limit: $limit) {
      id
      title
      coverArtUrl
      releaseYear
      artist {
        id
        name
      }
    }
    searchTracks(query: $query, limit: $limit) {
      id
      title
      durationMs
      formattedDuration
      streamUrl
      artist {
        id
        name
      }
      album {
        id
        title
        coverArtUrl
      }
    }
  }
`
